File: lib/page.dart
num _pxToInches(num px) => px / 96;

num _cmToInches(num cm) => cm / 2.54;

num _mmToInches(num mm) => _cmToInches(mm / 10);

class PaperFormat {
  static PaperFormat a3({required bool isPortrait}) =>
      PaperFormat.inches(width: 11.69, height: 16.54, isPortrait: isPortrait);
  static PaperFormat a4({required bool isPortrait}) =>
      PaperFormat.inches(width: 8.27, height: 11.69, isPortrait: isPortrait);
  static PaperFormat a5({required bool isPortrait}) =>
      PaperFormat.inches(width: 5.83, height: 8.27, isPortrait: isPortrait);
  static PaperFormat b4({required bool isPortrait}) =>
      PaperFormat.inches(width: 9.84, height: 13.90, isPortrait: isPortrait);
  static PaperFormat b5({required bool isPortrait}) =>
      PaperFormat.inches(width: 6.93, height: 9.84, isPortrait: isPortrait);
  static PaperFormat executive({required bool isPortrait}) =>
      PaperFormat.inches(width: 7.25, height: 10.5, isPortrait: isPortrait);
  static PaperFormat legal({required bool isPortrait}) =>
      PaperFormat.inches(width: 8.5, height: 14, isPortrait: isPortrait);
  static PaperFormat letter({required bool isPortrait}) =>
      PaperFormat.inches(width: 8.5, height: 11, isPortrait: isPortrait);
  static PaperFormat tabloid({required bool isPortrait}) =>
      PaperFormat.inches(width: 11, height: 17, isPortrait: isPortrait);

  const PaperFormat.inches({
    required this.width,
    required this.height,
    required this.isPortrait,
  });

  final num width, height;
  final bool isPortrait;

  PaperFormat.px({
    required int width,
    required int height,
    required this.isPortrait,
  })  : width = _pxToInches(width),
        height = _pxToInches(height);

  PaperFormat.cm({
    required num width,
    required num height,
    required this.isPortrait,
  })  : width = _cmToInches(width),
        height = _cmToInches(height);

  PaperFormat.mm({
    required num width,
    required num height,
    required this.isPortrait,
  })  : width = _mmToInches(width),
        height = _mmToInches(height);

  Map<String, num> toMap() {
    return {
      "width": isPortrait ? width : height,
      "height": isPortrait ? height : width,
    };
  }

  @override
  String toString() => 'PaperFormat.inches(width: $width, height: $height)';
}

class PdfMargins {
  num top, bottom, left, right;

  static final PdfMargins zero = PdfMargins.inches();

  PdfMargins({this.top = 0, this.bottom = 0, this.left = 0, this.right = 0});

  PdfMargins.inches({num? top, num? bottom, num? left, num? right})
      : top = top ?? 0,
        bottom = bottom ?? 0,
        left = left ?? 0,
        right = right ?? 0;

  /// [PdfMargins.px] then return in Inches
  factory PdfMargins.px({int top = 0, int bottom = 0, int left = 0, int right = 0}) {
    return PdfMargins.inches(
      top: _pxToInches(top),
      bottom: _pxToInches(bottom),
      left: _pxToInches(left),
      right: _pxToInches(right),
    );
  }

  /// [PdfMargins.cm] then return in Inches
  factory PdfMargins.cm({num top = 0, num bottom = 0, num left = 0, num right = 0}) {
    return PdfMargins.inches(
      top: _cmToInches(top),
      bottom: _cmToInches(bottom),
      left: _cmToInches(left),
      right: _cmToInches(right),
    );
  }

  /// [PdfMargins.mm] then return in Inches
  factory PdfMargins.mm({num top = 0, num bottom = 0, num left = 0, num right = 0}) {
    return PdfMargins.inches(
      top: _mmToInches(top),
      bottom: _mmToInches(bottom),
      left: _mmToInches(left),
      right: _mmToInches(right),
    );
  }

  Map<String, num> toMap() {
    return {
      "top": top,
      "bottom": bottom,
      "left": left,
      "right": right,
    };
  }

  @override
  String toString() => 'PdfMargins.inches(top: $top, bottom: $bottom, left: $left, right: $right)';
}

File: android/src/main/kotlin/android/print/Page.kt
package android.print

// ✅ CONVERSION FUNCTIONS: Helper functions for unit conversions
fun pxToInches(px: Double): Double = px / 96.0

fun inchToPx(d: Double): Double = d * 96

fun cmToInches(cm: Double): Double = pxToInches(cm / 37.8)

fun mmToInches(mm: Double): Double = cmToInches(mm / 10.0)

// ✅ PAPER FORMAT CLASS: Kotlin equivalent of your Dart PaperFormat
data class PaperFormat(
    val width: Double,
    val height: Double
) {
    companion object {
        // ✅ PREDEFINED PAPER SIZES: Static constants like your Dart version
        val LETTER = PaperFormat(width = 8.5, height = 11.0)
        val LEGAL = PaperFormat(width = 8.5, height = 14.0)
        val TABLOID = PaperFormat(width = 11.0, height = 17.0)
        val LEDGER = PaperFormat(width = 17.0, height = 11.0)
        val A0 = PaperFormat(width = 33.1, height = 46.8)
        val A1 = PaperFormat(width = 23.4, height = 33.1)
        val A2 = PaperFormat(width = 16.54, height = 23.4)
        val A3 = PaperFormat(width = 11.7, height = 16.54)
        val A4 = PaperFormat(width = 8.27, height = 11.7)
        val A5 = PaperFormat(width = 5.83, height = 8.27)
        val A6 = PaperFormat(width = 4.13, height = 5.83)

        // ✅ FROM STRING FACTORY: Equivalent to your Dart factory constructor
        fun fromString(value: String): PaperFormat {
            return when (value.lowercase()) {
                "letter" -> LETTER
                "legal" -> LEGAL
                "tabloid" -> TABLOID
                "ledger" -> LEDGER
                "a0" -> A0
                "a1" -> A1
                "a2" -> A2
                "a3" -> A3
                "a4" -> A4
                "a5" -> A5
                "a6" -> A6
                else -> A4 // Default to A4
            }
        }

        // ✅ FACTORY CONSTRUCTORS: Different unit constructors
        fun inches(width: Double, height: Double): PaperFormat {
            return PaperFormat(width = width, height = height)
        }

        fun px(width: Int, height: Int): PaperFormat {
            return PaperFormat(
                width = pxToInches(width.toDouble()),
                height = pxToInches(height.toDouble())
            )
        }

        fun cm(width: Double, height: Double): PaperFormat {
            return PaperFormat(
                width = cmToInches(width),
                height = cmToInches(height)
            )
        }

        fun mm(width: Double, height: Double): PaperFormat {
            return PaperFormat(
                width = mmToInches(width),
                height = mmToInches(height)
            )
        }
    }

    // ✅ TO MAP: Convert to map for easy serialization
    fun toMap(): Map<String, Double> {
        return mapOf(
            "width" to width,
            "height" to height
        )
    }

    // ✅ PIXELS: Get dimensions in pixels (96 DPI)
    val widthPixels: Int get() = (width * 96).toInt()
    val heightPixels: Int get() = (height * 96).toInt()

    // ✅ TO STRING: Readable string representation
    override fun toString(): String {
        return "PaperFormat.inches(width: $width, height: $height)"
    }
}

// ✅ PDF MARGINS CLASS: Kotlin equivalent of your Dart PdfMargins
data class PdfMargins(
    val top: Double = 0.0,
    val bottom: Double = 0.0,
    val left: Double = 0.0,
    val right: Double = 0.0
) {
    companion object {
        // ✅ ZERO MARGINS: Static constant
        val ZERO = PdfMargins()

        // ✅ FACTORY CONSTRUCTORS: Different unit constructors
        fun inches(top: Double = 0.0, bottom: Double = 0.0, left: Double = 0.0, right: Double = 0.0): PdfMargins {
            return PdfMargins(top = top, bottom = bottom, left = left, right = right)
        }

        fun px(top: Int = 0, bottom: Int = 0, left: Int = 0, right: Int = 0): PdfMargins {
            return PdfMargins(
                top = pxToInches(top.toDouble()),
                bottom = pxToInches(bottom.toDouble()),
                left = pxToInches(left.toDouble()),
                right = pxToInches(right.toDouble())
            )
        }

        fun cm(top: Double = 0.0, bottom: Double = 0.0, left: Double = 0.0, right: Double = 0.0): PdfMargins {
            return PdfMargins(
                top = cmToInches(top),
                bottom = cmToInches(bottom),
                left = cmToInches(left),
                right = cmToInches(right)
            )
        }

        fun mm(top: Double = 0.0, bottom: Double = 0.0, left: Double = 0.0, right: Double = 0.0): PdfMargins {
            return PdfMargins(
                top = mmToInches(top),
                bottom = mmToInches(bottom),
                left = mmToInches(left),
                right = mmToInches(right)
            )
        }
    }

    // ✅ TO MAP: Convert to map for easy serialization
    fun toMap(): Map<String, Double> {
        return mapOf(
            "top" to top,
            "bottom" to bottom,
            "left" to left,
            "right" to right
        )
    }

    // ✅ TO STRING: Readable string representation
    override fun toString(): String {
        return "PdfMargins.inches(top: $top, bottom: $bottom, left: $left, right: $right)"
    }
}